### PR TITLE
Add Ability to persist consoles queries locally. Provide correspondin…

### PR DESCRIPTION
**Pull request by YevheniiKoshevyi**
…g UI changes.

Signed-off-by: Your Real Name your.email@email.com

### CODE REVIEW

**Comment by ghost** [comment]
Good job!

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/cache/impl/HybrisConsoleInMemoryRegionsCache.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public class HybrisConsoleInMemoryRegionsCache implements HybrisConsoleRegionsCache {
+
+    private static final String SOLR = "SOLR";
+    private static final String FLEXIBLE_SEARCH = "FLEXIBLE_SEARCH";
+
+
+    private Map<String, Region> regions;
+
+    public HybrisConsoleInMemoryRegionsCache() {
+        this.regions = new HashMap<>(3);
```
Comment (ghost) [inline src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/cache/impl/HybrisConsoleInMemoryRegionsCache.java]: 2

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/ui/listeners/HybrisConsoleQueryPanelEventManager.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class HybrisConsoleQueryPanelEventManager {
+
+    private List<HybrisConsoleEventListener> listeners = new ArrayList<>();
+
+    public HybrisConsoleQueryPanelEventManager(Project project) {
```
Comment (ghost) [inline src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/ui/listeners/HybrisConsoleQueryPanelEventManager.java]: I think we can delete that constructor

File: common/src/com/intellij/idea/plugin/hybris/common/HybrisConstants.java [hunk 3]
```diff
@@ -101,6 +101,10 @@
     String DEFAULT_JAVADOC_ROOT_URL = "https://download.hybris.com/api/%s/commercesuite";
     String HYBRIS_6_0_PLUS_JAVADOC_ROOT_URL = "https://help.hybris.com/%s/api/commercesuite/index.html";
 
+    String QUERY_STORAGE_FOLDER_PATH = EXCLUDE_IDEA_DIRECTORY + "/consolestorage";
+    String SOLR_REGION_FILE_PATH = "/solr_region.json";
```
Comment (ghost) [inline common/src/com/intellij/idea/plugin/hybris/common/HybrisConstants.java]: Could we generate these file names depending from region name and not save here?

Comment (ghost) [reply common/src/com/intellij/idea/plugin/hybris/common/HybrisConstants.java]: regionname.toLowercase()+"_region.json"

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/pojo/RegionEntityMap.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.intellij.idea.plugin.hybris.tools.remote.console.persistence.pojo;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class RegionEntityMap<T> extends LinkedHashMap<String, RegionEntity<T>> {
```
Comment (ghost) [inline src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/pojo/RegionEntityMap.java]: Hm, maybe we need to rename it to RegionEntityFIFOCache

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/services/impl/DefaultConsolePersistenceService.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    private void loadEntityFromFile(
+        final RegionPersistenceService loadService,
+        final Path path,
+        final String regionName
+    ) throws IOException {
+        try {
+            loadService.loadRegionData(path, regionName);
+        } catch (IllegalArgumentException e) {
+            final String notificationTitle = e.getMessage();
+            final String notificationName = "Only 'IMPEX', 'FLEXIBLE_SEARCH' or 'SOLR' are allowable regions.";
```
Comment (ghost) [inline src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/services/impl/DefaultConsolePersistenceService.java]: ~IMPEX~

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/cache/impl/HybrisConsoleInMemoryRegionsCache.java (hunk 1 above)
Comment (YevheniiKoshevyi) [reply src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/cache/impl/HybrisConsoleInMemoryRegionsCache.java]: Sorry, forgot to change.

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/ui/listeners/HybrisConsoleQueryPanelEventManager.java (hunk 2 above)
Comment (YevheniiKoshevyi) [reply src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/ui/listeners/HybrisConsoleQueryPanelEventManager.java]: Yeah, you are right. It is redundant

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/pojo/RegionEntityMap.java (hunk 4 above)
Comment (YevheniiKoshevyi) [reply src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/pojo/RegionEntityMap.java]: Yes, RegionEntityFIFOCache sounds better

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/services/impl/DefaultConsolePersistenceService.java (hunk 5 above)
Comment (YevheniiKoshevyi) [reply src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/services/impl/DefaultConsolePersistenceService.java]: Got it

File: common/src/com/intellij/idea/plugin/hybris/common/HybrisConstants.java (hunk 3 above)
Comment (YevheniiKoshevyi) [reply common/src/com/intellij/idea/plugin/hybris/common/HybrisConstants.java]: I guess It makes sense

**Review by ghost — APPROVED** [review]

File: src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/ui/HybrisConsoleQueryPanel.kt [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+    private val saveButton = JButton(IconLoader.getIcon("/icons/menu-saveall.svg"))
+    private val loadButton = JButton(IconLoader.getIcon("/icons/upload.svg"))
+    private val removeButton = JButton(IconLoader.getIcon("/icons/delete.png"))
+
+    private val regionEntityService = RegionEntityService.getInstance(project)
+    private val regionService = RegionService.getInstance(project)
+    private val region = regionService.findOrCreate(region)
+
+    private var regionEntitiesComboBox: ComboBox<RegionEntity<*>>? = null
```
Comment (ghost) [inline src/com/intellij/idea/plugin/hybris/tools/remote/console/persistence/ui/HybrisConsoleQueryPanel.kt]: we could try:
```
private var regionEntitiesComboBox = ComboBox<RegionEntity<*>>(emptyArray(), 200)
....
 private fun addComponentsToPanel() {
...
 addComboBox()
}


 private fun addComboBox() {
....
 addRegionEntitiesToComboBox()
}

private inner class HybrisConsoleQueryPanelListener() : HybrisConsoleEventListener {
        override fun update() {
            addRegionEntitiesToComboBox()
        }
    }

    private fun addRegionEntitiesToComboBox() {
        if (regionEntitiesComboBox.itemCount == 0)
            regionEntitiesComboBox.model = DefaultComboBoxModel<RegionEntity<*>>(regionEntityService.getAll(this.region.name).values.toTypedArray())

    }

```

**Pull request merged**